File: portfolio/src/main/java/dev/exit/asset/portfolio/PortfolioSummaryService.java
package dev.exit.asset.portfolio;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.YearMonth;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import dev.exit.asset.domain.AssetItem;
import dev.exit.asset.domain.CashFlowTransaction;
import dev.exit.asset.domain.ExitGoal;
import dev.exit.asset.domain.Liability;
import dev.exit.asset.domain.TransactionType;

@Service
public class PortfolioSummaryService {

	private final AssetItemRepository assets;

	private final LiabilityRepository liabilities;

	private final CashFlowTransactionRepository transactions;

	private final ExitGoalRepository goals;

	public PortfolioSummaryService(AssetItemRepository assets, LiabilityRepository liabilities,
			CashFlowTransactionRepository transactions, ExitGoalRepository goals) {
		this.assets = assets;
		this.liabilities = liabilities;
		this.transactions = transactions;
		this.goals = goals;
	}

	@Transactional(readOnly = true)
	public PortfolioSummary summarize(String ownerId, YearMonth month) {
		BigDecimal totalAssets = assets.findByOwnerIdOrderByCreatedAtDesc(ownerId).stream()
			.map(AssetItem::getCurrentValue)
			.reduce(BigDecimal.ZERO, BigDecimal::add);
		BigDecimal totalLiabilities = liabilities.findByOwnerIdOrderByCreatedAtDesc(ownerId).stream()
			.map(Liability::getOutstandingBalance)
			.reduce(BigDecimal.ZERO, BigDecimal::add);
		BigDecimal netWorth = totalAssets.subtract(totalLiabilities);

		var monthlyTransactions = transactions.findByOwnerIdAndTransactedOnBetween(ownerId, month.atDay(1),
				month.atEndOfMonth());
		BigDecimal income = sumByType(monthlyTransactions, TransactionType.INCOME);
		BigDecimal expense = sumByType(monthlyTransactions, TransactionType.EXPENSE);
		BigDecimal savings = income.subtract(expense);
		BigDecimal savingsRate = income.signum() == 0 ? BigDecimal.ZERO
				: savings.multiply(BigDecimal.valueOf(100)).divide(income, 2, RoundingMode.HALF_UP);

		ExitGoal activeGoal = goals.findFirstByOwnerIdAndActiveTrueOrderByCreatedAtDesc(ownerId).orElse(null);
		BigDecimal goalProgress = activeGoal == null || activeGoal.getTargetNetWorth().signum() == 0 ? BigDecimal.ZERO
				: netWorth.multiply(BigDecimal.valueOf(100)).divide(activeGoal.getTargetNetWorth(), 2,
						RoundingMode.HALF_UP);
		Integer monthsToExit = activeGoal == null || savings.signum() <= 0 ? null
				: activeGoal.getTargetNetWorth().subtract(netWorth).max(BigDecimal.ZERO)
					.divide(savings, 0, RoundingMode.CEILING)
					.intValue();

		return new PortfolioSummary(totalAssets, totalLiabilities, netWorth, income, expense, savings, savingsRate,
				goalProgress, monthsToExit);
	}

	private BigDecimal sumByType(Iterable<CashFlowTransaction> items, TransactionType type) {
		BigDecimal total = BigDecimal.ZERO;
		for (CashFlowTransaction item : items) {
			if (item.getType() == type) {
				total = total.add(item.getAmount());
			}
		}
		return total;
	}

	public record PortfolioSummary(BigDecimal totalAssets, BigDecimal totalLiabilities, BigDecimal netWorth,
			BigDecimal monthlyIncome, BigDecimal monthlyExpense, BigDecimal monthlySavings, BigDecimal savingsRate,
			BigDecimal goalProgress, Integer monthsToExit) {
	}
}

File: portfolio/src/main/kotlin/dev/exit/asset/portfolio/PortfolioSummaryService.kt
package dev.exit.asset.portfolio

import dev.exit.asset.asset.AssetItem
import dev.exit.asset.asset.AssetItemRepository
import dev.exit.asset.cashflow.CashFlowTransaction
import dev.exit.asset.cashflow.CashFlowTransactionRepository
import dev.exit.asset.cashflow.TransactionType
import dev.exit.asset.goal.ExitGoalRepository
import dev.exit.asset.liability.Liability
import dev.exit.asset.liability.LiabilityRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.YearMonth

@Service
class PortfolioSummaryService(
    private val assets: AssetItemRepository,
    private val liabilities: LiabilityRepository,
    private val transactions: CashFlowTransactionRepository,
    private val goals: ExitGoalRepository,
) {

    @Transactional(readOnly = true)
    fun summarize(ownerId: String, month: YearMonth): PortfolioSummary {
        val totalAssets = assets.findByOwnerIdOrderByCreatedAtDesc(ownerId)
            .map(AssetItem::currentValue)
            .fold(BigDecimal.ZERO, BigDecimal::add)
        val totalLiabilities = liabilities.findByOwnerIdOrderByCreatedAtDesc(ownerId)
            .map(Liability::outstandingBalance)
            .fold(BigDecimal.ZERO, BigDecimal::add)
        val netWorth = totalAssets.subtract(totalLiabilities)

        val monthlyTransactions = transactions.findByOwnerIdAndTransactedOnBetween(
            ownerId,
            month.atDay(1),
            month.atEndOfMonth(),
        )
        val income = sumByType(monthlyTransactions, TransactionType.INCOME)
        val expense = sumByType(monthlyTransactions, TransactionType.EXPENSE)
        val savings = income.subtract(expense)
        val savingsRate = if (income.signum() == 0) {
            BigDecimal.ZERO
        } else {
            savings.multiply(BigDecimal.valueOf(100)).divide(income, 2, RoundingMode.HALF_UP)
        }

        val activeGoal = goals.findFirstByOwnerIdAndActiveTrueOrderByCreatedAtDesc(ownerId).orElse(null)
        val goalProgress = if (activeGoal == null || activeGoal.targetNetWorth.signum() == 0) {
            BigDecimal.ZERO
        } else {
            netWorth.multiply(BigDecimal.valueOf(100)).divide(activeGoal.targetNetWorth, 2, RoundingMode.HALF_UP)
        }
        val monthsToExit = if (activeGoal == null || savings.signum() <= 0) {
            null
        } else {
            activeGoal.targetNetWorth
                .subtract(netWorth)
                .max(BigDecimal.ZERO)
                .divide(savings, 0, RoundingMode.CEILING)
                .toInt()
        }

        return PortfolioSummary(
            totalAssets,
            totalLiabilities,
            netWorth,
            income,
            expense,
            savings,
            savingsRate,
            goalProgress,
            monthsToExit,
        )
    }

    private fun sumByType(items: Iterable<CashFlowTransaction>, type: TransactionType): BigDecimal =
        items
            .filter { it.type == type }
            .fold(BigDecimal.ZERO) { total, item -> total.add(item.amount) }

    data class PortfolioSummary(
        val totalAssets: BigDecimal,
        val totalLiabilities: BigDecimal,
        val netWorth: BigDecimal,
        val monthlyIncome: BigDecimal,
        val monthlyExpense: BigDecimal,
        val monthlySavings: BigDecimal,
        val savingsRate: BigDecimal,
        val goalProgress: BigDecimal,
        val monthsToExit: Int?,
    )
}
